optaplanner-examples/src/main/java/org/optaplanner/examples/investment/optional/score/InvestmentConstraintProvider.java

```diff
@@ -0,0 +1,61 @@
+package org.optaplanner.examples.investment.optional.score;
+
+import java.util.function.Function;
+
+import org.optaplanner.core.api.score.buildin.hardsoftlong.HardSoftLongScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintCollectors;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.Joiners;
+import org.optaplanner.examples.investment.domain.AssetClassAllocation;
+import org.optaplanner.examples.investment.domain.InvestmentParametrization;
+import org.optaplanner.examples.investment.domain.Region;
+import org.optaplanner.examples.investment.domain.Sector;
+
+public class InvestmentConstraintProvider implements ConstraintProvider {
+
+    private static final String CONSTRAINT_PACKAGE = "org.optaplanner.examples.investment.solver";
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        // TODO Auto-generated method stub
+        return new Constraint[]{
+                                assetsDeviationGreaterThanMaximumPenalty(constraintFactory),
+                                regionQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                sectorQuantityGreaterThanMaximumPenalty(constraintFactory),
+                                assetExpectedReturnReward(constraintFactory)
+        };
+    }
+
+    private Constraint assetsDeviationGreaterThanMaximumPenalty(ConstraintFactory cf) {
+        return cf.from(AssetClassAllocation.class)
+                 .join(AssetClassAllocation.class)
```

Comment:
`.fromUniquePair(AssetClassAllocation.class)`?

Reply:
Nope, duplicates are required for FULL_ASSERT to pass (apparently `a - a > 0` for investments)

Reply:
Would be nice to add a comment here then, explaining the situation.
I can already see some poor well-meaning soul "fixing" this unknowingly. :-)